segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/MetadataTransaction.java

```diff
@@ -0,0 +1,222 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.pravega.segmentstore.storage.metadata;
+
+import com.google.common.base.Preconditions;
+
+import lombok.Getter;
+import lombok.Setter;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Default implementation of the storage metadata transaction.
+ * All access to and modifications to the metadata the {@link ChunkMetadataStore} must be done through a transaction.
+ * This implementation delegates all calls to underlying {@link ChunkMetadataStore}.
+ * <div>
+ * A transaction is created by calling {@link ChunkMetadataStore#beginTransaction()}
+ * <ul>
+ * <li>Changes made to metadata inside a transaction are not visible until a transaction is committed using any overload of{@link MetadataTransaction#commit()}.</li>
+ * <li>Transaction is aborted automatically unless committed or when {@link MetadataTransaction#abort()} is called.</li>
```

Comment:
MetadataTransaction object just accumulates the changes and on calling commit it writes to underlying KV store.
If commit is not called the data is never written anywhere and simply discarded/thrown away. The java GC will take care of the rest.

Reply:
`MetadataTransaction `- keeps track of all KV pairs read by the transaction and keeps in memory copy of all the data. This allows isolation. Until transaction is committed , nothing is write to underlying store.

BaseMetadataStore - all reads and write to KV store go through this layer. So it always has complete view of all things read/modified/committed etc.
It actually accepts the transaction object and then writes it to underlying KV store. Every commit results in write to the store. This is critical data is not written until commit is called - until then all data is in memory only.

BaseMetadataStore also caches most recently committed records so that for frequent reads it doesn't have to read them again from KV store. There is bit of logic to make sure when we commit new records we also update the cached records. This layer also validates the commit to make sure it is not based invalid versions  etc. 

There are two optimizations
- lazy commit is when we update the record and use this updated copy in subsequent transactions but don't actually write to underlying KV store. Only valid use of this is when the length of the chunk is updated. Here we avoid making frequent writes to KV store, this critically depends on the fact that when we open a segment we check the actual length on LTS and update the in memory copy.

Second optimization is pinned records - which are never written to underlying KV store. These are only applicable for storage system segments. There is a separate SystemJournal to recreate the state of these segments.